### PR TITLE
#173 Export log to .csv

### PR DESCRIPTION
**Pull request by aszalos-roland**
# Export log to .csv implementation
A working solution for #173

## Description
Added 2 buttons at the top of CurveValuesView.

A Download all button:
image

A Download selected button:
image

The download selected button is only visible when there are rows selected.
These two buttons are disabled while the table is loading data.
Export is instantaneous except when there are more than 200k rows.

Also added a custom hook called useExport to make export functionality somewhat reusable, passing configuration props to this hook, one can define the file output.

## Type of change
- [ ] Bugfix
- [x] New feature (non-breaking change which adds functionality)
- [ ] Enhancement of existing functionality
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist:
- [ ] PR is related to an issue
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have used descriptive naming on components, functions and variables to avoid additional explanatory comments in the code
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings

## Further comments

Could consider adding export progress indicator when there are 200k+ rows.

### CODE REVIEW

File: Src/WitsmlExplorer.Frontend/components/ContentViews/table/tableParts.ts [hunk 1]
```diff
@@ -1,5 +1,9 @@
 import React from "react";
 
+export interface ExportableContentTableColumn<T> extends ContentTableColumn {
+  columnOf: T;
+}
```
Comment (aszalos-roland) [inline Src/WitsmlExplorer.Frontend/components/ContentViews/table/tableParts.ts:5]: anyone knows a good name for the class and property?

File: Src/WitsmlExplorer.Frontend/hooks/useExport.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
+export interface ExportOptions {
+  outputMimeType: string;
+  fileExtension: string;
+  separator: string;
+  newLineCharacter: string;
+  omitSpecialCharactersFromFilename?: boolean;
+  appendTimeStamp?: boolean;
+}
+interface ExportReturn {
```
Comment (aszalos-roland) [inline Src/WitsmlExplorer.Frontend/hooks/useExport.ts]: anyone knows a good name for this interface?

**Comment by HakonHesselberg** [comment]
Button looks like this while data is being fetched:
image

Should maybe have a more clear "disabled" looking button?

**Comment by HakonHesselberg** [comment]
Possible alternative:
Moving the download functionality to the context menu.
Removing the "Download all" functionality as there is an option to check all rows.

These are just suggestions, not sure what is the best solution.

File: Src/WitsmlExplorer.Frontend/components/ContentViews/CurveValuesView.tsx [hunk 3]
```diff
@@ -129,18 +157,43 @@ export const CurveValuesView = (): React.ReactElement => {
 
   return (
     <Container>
+      {Boolean(tableData.length) && (
+        <StyledGrid container spacing={1}>
```
Comment (HakonHesselberg) [inline Src/WitsmlExplorer.Frontend/components/ContentViews/CurveValuesView.tsx]: Consider more specific name like `ExportButtonGrid` or similar?

**Review by HakonHesselberg — APPROVED** [review]
Works on my machine, both selected and download all

File: Src/WitsmlExplorer.Frontend/components/ContentViews/CurveValuesView.tsx (hunk 3 above)
Comment (aszalos-roland) [reply Src/WitsmlExplorer.Frontend/components/ContentViews/CurveValuesView.tsx]: good idea

File: Src/WitsmlExplorer.Frontend/hooks/useExport.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+function useExport(props: ExportOptions): ExportReturn {
+  const exportData = (fileName: string, header: string, data: string) => {
+    const link = document.createElement("a");
+    link.href = window.URL.createObjectURL(
+      new Blob([header, props.newLineCharacter, data], {
+        type: props.outputMimeType
+      })
+    );
+    link.download = props.omitSpecialCharactersFromFilename
+      ? `${omitSpecialCharacters(fileName)}${appendTimeStamp(props.appendTimeStamp)}${props.fileExtension}`
```
Comment (marmid74) [inline Src/WitsmlExplorer.Frontend/hooks/useExport.ts]: Updated the specification in userstory #173 for timestamp. The requirement is a human readable format, and the filename should be in following format "Wellbore name"-"Log name"-"YYYY"-"MM"-"DD"T"HH-MM-SS.csv

File: Src/WitsmlExplorer.Frontend/hooks/useExport.ts [hunk 5]
```diff
@@ -0,0 +1,41 @@
+export interface ExportOptions {
```
Comment (marmid74) [inline Src/WitsmlExplorer.Frontend/hooks/useExport.ts]: as only one property is optional, "ExportOptions" is not spot on. Perhaps ExportProperties?

Comment (aszalos-roland) [reply Src/WitsmlExplorer.Frontend/hooks/useExport.ts]: okay, I renamed it

File: Src/WitsmlExplorer.Frontend/hooks/useExport.ts (hunk 4 above)
Comment (aszalos-roland) [reply Src/WitsmlExplorer.Frontend/hooks/useExport.ts]: updated the format in the latest commit

Comment (marmid74) [reply Src/WitsmlExplorer.Frontend/hooks/useExport.ts]: > Button looks like this while data is being fetched:
> image
> 
> Should maybe have a more clear "disabled" looking button?

Please have a look at UserCredentialsModal.tsx and ModalDialog.txs for example on disabled button.
image

File: Src/WitsmlExplorer.Frontend/hooks/useExport.ts (hunk 2 above)
Comment (marmid74) [reply Src/WitsmlExplorer.Frontend/hooks/useExport.ts]: This was a difficult one...

File: Src/WitsmlExplorer.Frontend/hooks/useExport.ts [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+  outputMimeType: string;
+  fileExtension: string;
+  separator: string;
+  newLineCharacter: string;
+  omitSpecialCharactersFromFilename?: boolean;
+  appendDateTime?: boolean;
+}
+interface ExportReturn {
+  exportData: (fileName: string, header: string, data: string) => void;
+  options: ExportProperties;
```
Comment (marmid74) [inline Src/WitsmlExplorer.Frontend/hooks/useExport.ts]: rename options  to properties ?

Comment (aszalos-roland) [reply Src/WitsmlExplorer.Frontend/hooks/useExport.ts]: forgot that, good point 👍

File: Src/WitsmlExplorer.Frontend/components/ContentViews/CurveValuesView.tsx [hunk 7]
```diff
@@ -1,22 +1,49 @@
...
+    setSelectedRows(orderBy([...rows.map((row) => row as CurveValueRow)], sortedColumn, sortOrder));
+  }, []);
+
+  const exportAll = useCallback(() => {
+    const exportColumns = columns.map((column) => `${column.columnOf.mnemonic}[${column.columnOf.unit}]`).join(exportOptions.separator);
+    const data = tableData.map((row) => columns.map((col) => row[col.columnOf.mnemonic] as string).join(exportOptions.separator)).join(exportOptions.newLineCharacter);
+    exportData(`${selectedWellbore.name}-${selectedLog.name}`, exportColumns, data);
+  }, [columns, tableData]);
+
+  const exportSelected = useCallback(() => {
```
Comment (marmid74) [inline Src/WitsmlExplorer.Frontend/components/ContentViews/CurveValuesView.tsx]: rename to exportSelectedDataPoints?

File: Src/WitsmlExplorer.Frontend/components/ContentViews/CurveValuesView.tsx [hunk 8]
```diff
@@ -1,22 +1,49 @@
...
+    separator: ",",
+    omitSpecialCharactersFromFilename: true,
+    appendDateTime: true
+  });
+
+  const rowSelectionCallback = useCallback((rows: ContentTableRow[], sortOrder: Order, sortedColumn: string) => {
+    setSelectedRows(orderBy([...rows.map((row) => row as CurveValueRow)], sortedColumn, sortOrder));
+  }, []);
+
+  const exportAll = useCallback(() => {
```
Comment (marmid74) [inline Src/WitsmlExplorer.Frontend/components/ContentViews/CurveValuesView.tsx]: rename to exportSelectedIndexRange?

File: Src/WitsmlExplorer.Frontend/hooks/useExport.ts (hunk 2 above)
Comment (marmid74) [reply Src/WitsmlExplorer.Frontend/hooks/useExport.ts]: What about ExportObject ?

Comment (aszalos-roland) [reply Src/WitsmlExplorer.Frontend/hooks/useExport.ts]: sounds good

File: Src/WitsmlExplorer.Frontend/hooks/useExport.ts (hunk 4 above)
Comment (aszalos-roland) [reply Src/WitsmlExplorer.Frontend/hooks/useExport.ts]: updated the button in the latest commit

File: Src/WitsmlExplorer.Frontend/hooks/useExport.ts (hunk 6 above)
Comment (aszalos-roland) [reply Src/WitsmlExplorer.Frontend/hooks/useExport.ts]: renamed in latest commit

File: Src/WitsmlExplorer.Frontend/components/ContentViews/CurveValuesView.tsx (hunk 7 above)
Comment (aszalos-roland) [reply Src/WitsmlExplorer.Frontend/components/ContentViews/CurveValuesView.tsx]: in latest commit

File: Src/WitsmlExplorer.Frontend/components/ContentViews/CurveValuesView.tsx (hunk 8 above)
Comment (aszalos-roland) [reply Src/WitsmlExplorer.Frontend/components/ContentViews/CurveValuesView.tsx]: in latest commit

**Comment by aszalos-roland** [comment]
> Button looks like this while data is being fetched:
> image
> 
> Should maybe have a more clear "disabled" looking button?

changed the behaviour in the latest commit

File: Src/WitsmlExplorer.Frontend/hooks/useExport.ts (hunk 2 above)
Comment (aszalos-roland) [reply Src/WitsmlExplorer.Frontend/hooks/useExport.ts]: renamed in latest commit

**Review by marmid74 — APPROVED** [review]
Approves, awaits the branch update prior merge.

**Comment by aszalos-roland** [comment]
> Approves, awaits the branch update prior merge.

updated the branch 👍

**Pull request merged**